src/Umbraco.Web/Compose/NestedContentPropertyComponent.cs

```diff
@@ -0,0 +1,128 @@
+
+                    // Remove keys from edited/draft value & any nested NC's
+                    var updatedEditedVal = CreateNestedContentKeys(cultureVal.EditedValue?.ToString(), onlyMissingKeys);
+                    cultureVal.EditedValue = updatedEditedVal;
+                }
+            }
+        }
+
+        private string CreateNestedContentKeys(string rawJson, bool onlyMissingKeys)
+        {
+            if (string.IsNullOrWhiteSpace(rawJson))
+                return rawJson;
+
+            // Parse JSON
+            var ncJson = JToken.Parse(rawJson);
+
+            // NC prop contains one or more items/rows of things
+            foreach (var nestedContentItem in ncJson.Children<JObject>())
+            {
+                // If saving/publishing - we only generate keys for NC items that are missing
+                if (onlyMissingKeys)
+                {
+                    var ncKeyProp = nestedContentItem.Properties().SingleOrDefault(x => x.Name.InvariantEquals("key"));
+                    if (ncKeyProp == null)
+                    {
+                        nestedContentItem.Properties().Append(new JProperty("key", Guid.NewGuid().ToString()));
+                    }
+                }
+
+                foreach (var ncItemProp in nestedContentItem.Properties())
+                {
+                    if (onlyMissingKeys == false)
+                    {
+                        // Only when copying a node - we generate new keys for all NC items
+                        if (ncItemProp.Name.InvariantEquals("key"))
+                            ncItemProp.Value = Guid.NewGuid().ToString();
+                    }
+
+                    // No need to check this property for JSON - as this is a JSON prop we know
+                    // That only contains the string of the doctype alias used as the NC item
+                    if (ncItemProp.Name == NestedContentPropertyEditor.ContentTypeAliasPropertyKey)
+                        continue;
+
+                    // As we don't know what properties in the JSON may contain other complex editors or deep nested NC
+                    // We are detecting if its value stores JSON to help filter the list AND that in its JSON it has ncContentTypeAlias prop
+                    var ncItemPropVal = ncItemProp.Value?.ToString();
+
+                    if (ncItemPropVal.DetectIsJson())
+                    {
+                        // Parse the nested JSON (complex editor)
+                        var complexEditorJson = JToken.Parse(ncItemPropVal);
+
+                        // Verify the complex editor is nested content (Will have one or more ncContentTypeAlias)
+                        // One for each NC item/row
+                        var hasNestedContentJsonProp = complexEditorJson.SelectTokens($"$..['{NestedContentPropertyEditor.ContentTypeAliasPropertyKey}']", false);
+                        if (hasNestedContentJsonProp.Count() > 0)
```

Comment:
This check will work if we have NC nested within another, but if we are currently in an editor that isn't NC that is within NC this won't work. Where are the unit tests? I don't understand how you are testing this stuff without that, it would be such a time consumer to re-test this over and over in real life plus you will never be sure you are covering all cases. It will be much easier to develop this against tests!  So this still wont work if you have NC -> some other editor -> NC especially when the sub json is escaped.

This is not doing what we talked about in our last call. If this is false, then we know the current JSON doesn't contain a direct reference to NC but we don't know if theres a sub-reference. What you had before where you just do a brute force string check if it contains `NestedContentPropertyEditor.ContentTypeAliasPropertyKey` is correct because we have to deal with escaped JSON. So do the string.invariantcontains check, if that is true, then you know somewhere nested within this json there's probably an NC, though it could also just be a string of that alias someone typed in, but at least at that stage you know you need to recurse further. I think this is just gonna be faster if i re-do this for you.